toolchain/check/deduce.cpp

```diff
@@ -507,8 +510,60 @@ auto DeductionContext::CheckDeductionIsComplete() -> bool {
       }
       return false;
     }
-  }
 
+    // If the binding is symbolic it can refer to other earlier bindings in the
+    // same generic, or from an enclosing specific. Substitute to replace those
+    // and get a non-symbolic type in order for us to know the final type that
+    // the argument needs to be converted to.
+    //
+    // Note that when typechecking a checked generic, the arguments can
+    // still be symbolic, so the substitution would also be symbolic. We are
+    // unable to get the final type for symbolic bindings until deducing with
+    // non-symbolic arguments.
+    //
+    // TODO: If arguments of different values, but that _convert to_ the same
+    // value, are deduced for the same symbolic binding, then we will fail
+    // typechecking in Deduce() with conflicting types via the
+    // `DeductionInconsistent` diagnostic. If we defer that check until after
+    // all conversions are done (after the code below) then we won't diagnose
+    // that incorrectly.
+    auto arg_type_id = context().insts().Get(deduced_arg_id).type_id();
+    auto binding_type_id = context().insts().Get(binding_id).type_id();
+    if (!arg_type_id.AsConstantId().is_symbolic() &&
+        binding_type_id.AsConstantId().is_symbolic()) {
+      auto param_type_const_id = SubstConstant(
+          context(), binding_type_id.AsConstantId(), substitutions_);
+      CARBON_CHECK(param_type_const_id.has_value());
+      CARBON_CHECK(!param_type_const_id.is_symbolic());
+      binding_type_id = context().GetTypeIdForTypeConstant(param_type_const_id);
+
+      // TODO: Suppress diagnostics here if `diagnose_` is false.
+      DiagnosticAnnotationScope annotate_diagnostics(
+          &context().emitter(),
+          [&](auto& builder) { NoteInitializingParam(binding_id, builder); });
+      auto converted_arg_id = ConvertToValueOfType(
+          context(), loc_id_, deduced_arg_id, binding_type_id);
+      // Replace the deduced arg with its value converted to the parameter
+      // type. The conversion of the argument type must produce a constant value
+      // to be used in deduction.
+      if (context().constant_values().Get(converted_arg_id).is_constant()) {
+        deduced_arg_id = converted_arg_id;
```

Comment:
One question here. We can't use the runtime value in the else case. But just leaving it as the original value - that would perhaps allow runtime conversions for a deduced generic type. Do we want to prevent that by setting the deduced arg to the Error instruction in the else case below?

Reply:
Yes, setting the argument to Error after issuing the error makes sense to me.

Reply:
Ok, thanks done.